script/lint.js

```diff
@@ -0,0 +1,196 @@
+  ['brightray', 'browser', 'mac', 'notification_center_delegate.h'],
+  ['brightray', 'browser', 'win', 'notification_presenter_win7.h'],
+  ['brightray', 'browser', 'win', 'win32_desktop_notifications', 'common.h'],
+  ['brightray', 'browser', 'win', 'win32_desktop_notifications',
+    'desktop_notification_controller.cc'],
+  ['brightray', 'browser', 'win', 'win32_desktop_notifications',
+    'desktop_notification_controller.h'],
+  ['brightray', 'browser', 'win', 'win32_desktop_notifications', 'toast.h'],
+  ['brightray', 'browser', 'win', 'win32_notification.h']
+].map(tokens => path.join(SOURCE_ROOT, ...tokens)))
+
+const LINTERS = [
+  {
+    key: 'c++',
+    roots: ['atom', 'brightray'].map(x => path.join(SOURCE_ROOT, x)),
+    test: filename => filename.endsWith('.cc') || filename.endsWith('.h'),
+    run: async (filenames) => {
+      childProcess.execFile('cpplint.py', filenames, {}, (error, stdout, stderr) => {
+        // cpplint writes warnings, errors, AND status messages to stderr.
+        // prune out the status messages:
+        for (const line of stderr.split(/[\r\n]+/)) {
+          if (line.length && !line.startsWith('Done processing ') && line !== 'Total errors found: 0') {
+            console.warn(line)
+          }
+        }
+        if (error || stderr.includes('Command failed')) {
+          process.exit(1)
+        }
+      })
+    }
+  }, {
+    key: 'python',
+    roots: ['script'].map(x => path.join(SOURCE_ROOT, x)),
+    test: filename => filename.endsWith('.py'),
+    run: async (filenames) => {
+      const rcfile = path.normalize(path.join(SOURCE_ROOT, '..', 'third_party', 'depot_tools', 'pylintrc'))
+      const args = ['--rcfile=' + rcfile, ...filenames]
+      const env = Object.assign({PYTHONPATH: path.join(SOURCE_ROOT, 'script')}, process.env)
+      childProcess.execFile('pylint.py', args, { env }, (unused, stdout, stderr) => {
+        if (!stdout.length && !stderr.length) { // pylint is quiet on success
+          return
+        }
+        console.warn(stdout, stderr)
+        process.exit(1)
+      })
+    }
+  }, {
+    key: 'javascript',
+    roots: ['lib', 'script'].map(x => path.join(SOURCE_ROOT, x)),
+    test: filename => filename.endsWith('.js'),
+    run: async (filenames) => {
+      const cmd = path.join(SOURCE_ROOT, 'node_modules', '.bin', 'standard')
+      childProcess.execFile(cmd, filenames, { cwd: SOURCE_ROOT }, (error, stdout, stderr) => {
+        if (!error) { // standard is quiet on success
+          return
+        }
+        console.warn(stdout)
+        process.exit(1)
+      })
+    }
+  }, {
+    key: 'javascript',
+    roots: ['spec'].map(x => path.join(SOURCE_ROOT, x)),
+    test: filename => filename.endsWith('.js'),
+    run: async (filenames) => {
+      const cmd = path.join(SOURCE_ROOT, 'node_modules', '.bin', 'standard')
+      childProcess.execFile(cmd, filenames, { cwd: path.join(SOURCE_ROOT, 'spec') }, (error, stdout, stderr) => {
```

Comment:
We're going to lose output doing this (only logging stdout / not streaming the std{out/err} streams in the correct order.

I'd much rather use logic that looks somthing like this

```js
const child = childProcess.spawn(cmd, args, { cwd: '', stdio: 'inherit' });
const exitCode = await new Promise(r => child.on('exit', r));
if (exitCode !== 0) process.exit(1)
```

Reply:
This is a good idea. Fixed.

Actually `cpplint.py` seems to be the outlier here -- the others can all be run with `spawnSync` with `inherit` and just check the exit code. So the latest revision is simpler still.